### DISCUSSION TITLE
语法问题

### DISCUSSION THREAD

**Discussion by Yawningx** [arcsysu/SYsU-lang#82] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/74655243/228724124-b2c9ca45-dbc8-4907-bd2b-6a85eb7894fa.png)
上图语法在测试以下样例时不能正常输出，报段错误：
```
int a;
int main(){
	a=10;
	return 0;
}

```
然而如果不用Functype，而是枚举T_INT或T_VOID，就可以正常输出，请问是我原本的语法有误吗

**Comment by wufeng15226** [comment]
可能是存在移入规约冲突，可以关注编译时的 warning

**Comment by feifeishenFFS** [comment]
我也遇见了这个问题，全局变量里面，Decl或者FuncDef的开头如果用Btype会报段错误，用T_INT才不会。

**Reply by Yawningx** [reply]
一模一样，请问你找到问题原因了吗😭

**Reply by feifeishenFFS** [reply]
暂时没发现出来，我只能拆开写着先了...
`note: rerun with option '-Wcounterexamples' to generate conflict counterexamples`
编译的时候有提示这个，暂时也不知道怎么搞冲突的问题

**Reply by wufeng15226** [reply]
你们可以在编译的时候加上这个选项看看提供的例子，理解冲突，再修改语法

**Comment by LjwNautilus** [comment, marked as answer]
确实是文法发生了 reduce-reduce conflict ，FuncType 和 BType 都可以直接推出 T_INT ，而且 T_INT 后面都可以接 T_IDENTIFIER ，bison 只看一个 symbol 不知道是规约成函数定义 FuncDef 还是变量声明 VarDecl 。报段错误是因为 bison 自动忽略了 FuncType 和 BType 中最靠后的规则，导致 parser 原来可以识别的很多结构都无法识别。

```yacc
 // A wrong example
FuncDef: 
  FuncType T_IDENTIFIER T_L_PAREN T_R_PAREN Block {
    $$ = AddTreeNode($5, "FunctionDecl", $2->name);
    delete $2;
  }
  ;

VarDecl:
  BType VarDef T_SEMI {
    $$ = AddTreeNode($2, "VarDecl");
  }
  ;

VarDef:
  T_IDENTIFIER {
    delete $1;
  }
  | T_IDENTIFIER T_EQUAL InitVal {
    $$ = $3;
    delete $1;
  }
  ;

FuncType: T_VOID | T_INT;
BType: T_INT | T_CHAR | T_LONG T_LONG;
```

编译时遇到 conflict 会报 warning 但信息很少，可以直接用 bison 加上 `-Wcounterexamples` 编译选项查看发生冲突的例子。
```bash
bison /workspace/SYsU-lang/parser/parser.y -Wcounterexamples
```
![image](https://user-images.githubusercontent.com/80296250/229751086-1584b522-d37f-4ffe-acfa-8133406aed56.png)

适当的枚举确实可以解决该冲突，我的方法是直接放弃使用 FuncType 。
```yacc
FuncDef: 
  T_VOID T_IDENTIFIER T_L_PAREN T_R_PAREN Block {
    $$ = AddTreeNode($5, "FunctionDecl", $2->name);
    delete $2;  // deletes node of the identifier to avoid being printed.
  }
  | BType T_IDENTIFIER T_L_PAREN T_R_PAREN Block {
    $$ = AddTreeNode($5, "FunctionDecl", $2->name);
    delete $2;  // deletes node of the identifier to avoid being printed.
  }
  ;
```

**Reply by Yawningx** [reply]
有理 给你点了👍
